server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -78,10 +78,12 @@ public ParsedDocument parse(SourceToParse source) throws MapperParsingException
 
     public void validate(IndexSettings settings, boolean checkLimits) {
         this.mapping().validate(this.mappingLookup);
-        if (settings.getIndexMetadata().isRoutingPartitionedIndex()) {
-            if (routingFieldMapper().required() == false) {
-                throw new IllegalArgumentException("mapping type [" + type() + "] must have routing "
-                    + "required for partitioned index [" + settings.getIndex().getName() + "]");
+        if (this.mapping().toString().equals("{\"_doc\":{\"properties\":{\"@timestamp\":{\"type\":\"date\"}}}}") == false) {
```

Comment:
What situation is this comparison of the string version of the mapping supposed to handle?

Reply:
in org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:1027 line : 
mappings.add(0, new CompressedXContent(wrapMappingsIfNecessary(DEFAULT_TIMESTAMP_MAPPING, xContentRegistry)));
 
datastream will add additonal DEFAULT_TIMESTAMP_MAPPING to the first position in mappings list, and 
in org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:1232 
for (CompressedXContent mapping : mappings) {
                        mapperService.merge(MapperService.SINGLE_MAPPING_NAME, mapping, MergeReason.INDEX_TEMPLATE);
                    }
it will merge(validate) one by one in mapping, for the first DEFAULT_TIMESTAMP_MAPPING, if user set routing_partition_size greater than 1 in settings of index, it will certainly fail. So I exclude the DEFAULT_TIMESTAMP_MAPPING in the validate.

Maybe we should only merge but not validate all mappings to get a final mapping, then we validate the final mapping ?

Reply:
Instead of this check, could we add another constant like `MetadataIndexTemplateService.DEFAULT_TIMESTAMP_MAPPING` that included the `_routing` field for data streams with custom routing enabled? That way, the data stream indices would have a routing metafield and the existing check here would still work.

@martijnvg, please let me know if I've mischaracterized your suggestion here.